File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def replaceChar(word, index):
    s = list(word)
    s[index] = 'X'
    return "".join(s)

def solution(begin, target, words):
    if target not in words:
        return 0
        
    wordList = {}
    for word in words:
        wordList[word] = 0
    
    def bfs():
        queue = deque([begin])
        wordList[begin] = 0
        
        while queue:
            cur = queue.popleft()
            
            for word in words:
                if word == begin:
                    continue
                
                for i in range(len(cur)):
                    if (replaceChar(cur, i) != replaceChar(word, i)):
                        continue
                    
                    if wordList[word] == 0 or wordList[word] > wordList[cur] + 1:
                        wordList[word] = wordList[cur] + 1     
                        queue.append(word)
                        
    bfs()

    answer = wordList[target]
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words) {
	let N = begin.length;

	let visited = {};
	for (let word of words) {
		visited[word] = false;
	}

	let isChanged = false;
	let answer = 1e9;
	function DFS(depth, word) {
		if (word === target) {
			answer = Math.min(answer, depth);
			isChanged = true;
			return;
		} else {
			for (let i = 0; i < N; i++) {
				let A = word.slice(0, i) + word.slice(i + 1);

				for (let x of words) {
					let B = x.slice(0, i) + x.slice(i + 1);

					if (A === B) {
						if (!visited[x]) {
							visited[x] = true;
							DFS(depth + 1, x);
							visited[x] = false;
						}
					}
				}
			}
		}
	}

	DFS(0, begin);

	return isChanged ? answer : 0;
}